### ISSUE TITLE
File Path return Null and File Option Detected as Unknown

### ISSUE THREAD

**Issue by JosephSanjaya** [javakam/FileOperator#21]
# Problem

I used this library for my app support chat system to send file and image, when i tried to send image theres no problem, but when i selecting files, FileType return is Unknown and the path is null. (File that i selected is Excel) i already define FIleOption and Mime. is my implementation is right? 

## Implementation

```kotlin
fun Fragment.chooseFile(requestCode: Int, callback: FileSelectCallBack) : FileSelector {
    val optionsWord = FileSelectOptions().apply {
        fileType = FileType.WORD
        singleFileMaxSize = Constant.BAREKSA_MAX_IMAGE_UPLODED.toLong()
        singleFileMaxSizeTip = StringUtils.getString(R.string.label_tv_bareksa_max_img_msg)
        allFilesMaxSize = Constant.BAREKSA_MAX_IMAGE_UPLODED.toLong()
    }
    val optionsExcel = FileSelectOptions().apply {
        fileType = FileType.EXCEL
        singleFileMaxSize = Constant.BAREKSA_MAX_IMAGE_UPLODED.toLong()
        singleFileMaxSizeTip = StringUtils.getString(R.string.label_tv_bareksa_max_img_msg)
    }
    val optionsPDF = FileSelectOptions().apply {
        fileType = FileType.PDF
        singleFileMaxSize = Constant.BAREKSA_MAX_IMAGE_UPLODED.toLong()
        singleFileMaxSizeTip = StringUtils.getString(R.string.label_tv_bareksa_max_img_msg)
    }
    val optionsPPT = FileSelectOptions().apply {
        fileType = FileType.PPT
        singleFileMaxSize = Constant.BAREKSA_MAX_IMAGE_UPLODED.toLong()
        singleFileMaxSizeTip = StringUtils.getString(R.string.label_tv_bareksa_max_img_msg)
    }
    val optionsTXT = FileSelectOptions().apply {
        fileType = FileType.TXT
        singleFileMaxSize = Constant.BAREKSA_MAX_IMAGE_UPLODED.toLong()
        singleFileMaxSizeTip = StringUtils.getString(R.string.label_tv_bareksa_max_img_msg)
    }
    val optionsZIP = FileSelectOptions().apply {
        fileType = FileType.ZIP
        singleFileMaxSize = Constant.BAREKSA_MAX_IMAGE_UPLODED.toLong()
        singleFileMaxSizeTip = StringUtils.getString(R.string.label_tv_bareksa_max_img_msg)
    }
    return FileSelector
            .with(this)
            .setRequestCode(requestCode)
            .setTypeMismatchTip("Tipe file tidak didukung")
            .setMinCount(1, "Pilih minimal 1 file!")
            .setMaxCount(1, "Pilih minimal 1 file!")
            .setOverLimitStrategy(FileGlobal.OVER_LIMIT_EXCEPT_ALL)
            .setMimeTypes(
                    "application/msword",
                    "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
                    "application/vnd.ms-powerpoint",
                    "application/vnd.openxmlformats-officedocument.presentationml.presentation",
                    "application/text/plain",
                    "application/vnd.ms-excel",
                    "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
                    "application/zip",
                    "application/pdf"
            )
            .applyOptions(optionsExcel, optionsPDF, optionsPPT, optionsTXT, optionsWord, optionsZIP)
            .filter(object : FileSelectCondition {
                override fun accept(fileType: IFileType, uri: Uri?): Boolean {
                    return if(uri != null && !uri.path.isNullOrBlank()) {
                        when(fileType) {
                            FileType.WORD, FileType.EXCEL, FileType.PDF,FileType.PPT, FileType.TXT, FileType.ZIP -> true
                            else -> false
                        }
                    } else false
                }
            })
            .callback(callback)
            .choose()
}
```
## Tested On
Google Pixel 4 (Android 11)

**Comment by javakam** [comment]
@JosephSanjaya  v1.3.6 我刚才用了你的代码测试,  在 Google Pixel 4 (Android 11) 模拟器上没有出现返回`UNKOWN`的情形,  `UNKOWN`是在选择了限定类型外的其它文件时才会出现

I just used your code to test, and there is no return to `UNKOWN` on the Google Pixel 4 (Android 11) emulator. `UNKOWN` will only appear when other files other than the limited type are selected.

测试文件路径 (Test file path)

![image](https://user-images.githubusercontent.com/15343661/108036763-9a776d00-7073-11eb-80f1-870e80e37b35.png)

运行结果 (Operation result)

![image](https://user-images.githubusercontent.com/15343661/108036442-318ff500-7073-11eb-9ba4-c16d7edf9bd0.png)

**Comment by javakam** [comment]
@JosephSanjaya  v1.3.6 我刚才用了你的代码测试,  在 Google Pixel 4 (Android 11) 模拟器上没有出现返回`UNKOWN`的情形,  `UNKOWN`是在选择了限定类型外的其它文件时才会出现

I just used your code to test, and there is no return to `UNKOWN` on the Google Pixel 4 (Android 11) emulator. `UNKOWN` will only appear when other files other than the limited type are selected.

测试文件路径 (Test file path)

![image](https://user-images.githubusercontent.com/15343661/108038234-90566e00-7075-11eb-9e1a-ba554307416c.png)

运行结果 (Operation result)

![image](https://user-images.githubusercontent.com/15343661/108036442-318ff500-7073-11eb-9ba4-c16d7edf9bd0.png)

**Comment by JosephSanjaya** [comment]
I Use 2 type of FileSelector 1 for Image and 1 for File but use 1 FileSelectCallBack, is it the source of problem?

this is My Implementation 

## Image Selector

```kotlin
fun Fragment.chooseImage(requestCode: Int, callback: FileSelectCallBack) : FileSelector {
    val optionsImage = FileSelectOptions().apply {
        fileType = FileType.IMAGE
        fileTypeMismatchTip = "Tipe file tidak didukung"
        minCount = 1
        maxCount = 1
        minCountTip = "Pilih minimal 1 gambar!"
        maxCountTip = "Pilih minimal 1 gambar!"
        fileCondition = object : FileSelectCondition {
            override fun accept(fileType: IFileType, uri: Uri?): Boolean {
                return (fileType == FileType.IMAGE && uri != null && !uri.path.isNullOrBlank() && !FileUtils.isGif(uri))
            }
        }
    }
    return FileSelector
            .with(this)
            .setRequestCode(requestCode)
            .setTypeMismatchTip("Tipe file tidak didukung")
            .setMinCount(1, "Pilih minimal 1 gambar!")
            .setMaxCount(1, "Pilih minimal 1 gambar!")
            .setOverLimitStrategy(FileGlobal.OVER_LIMIT_EXCEPT_ALL)
            .setMimeTypes("image/*")
            .applyOptions(optionsImage)
            .filter(object : FileSelectCondition {
                override fun accept(fileType: IFileType, uri: Uri?): Boolean {
                    return when (fileType) {
                        FileType.IMAGE -> (uri != null && !uri.path.isNullOrBlank() && !FileUtils.isGif(uri))
                        FileType.VIDEO -> false
                        FileType.AUDIO -> false
                        else -> false
                    }
                }
            })
            .callback(callback)
            .choose()
}
```

File Selector is the same as code on last comment

## Fragment Implementation

```kotlin
class FinChatFragment : FileSelectCallBack {

    companion object {
        private const val REQUEST_CHOOSE_IMAGE = 1245154
        private const val REQUEST_CHOOSE_FILE = 124151
    }
    
    private var mFileSelector: FileSelector? = null
    private var mFileSelectorRequest = REQUEST_CHOOSE_IMAGE
    
    private val bottomDialog by lazy {
        mContext.makeCustomViewDialog(ViewBottomMenuBinding::inflate, isTransparent = true, isBottomDialog = true).apply {
            first.tvFirst.setOnClickListener {
                mFileSelectorRequest = REQUEST_CHOOSE_IMAGE
                mFileSelector = chooseImage(mFileSelectorRequest, this@FinChatFragment)
                second.dismiss()
            }
            first.tvSecond.setOnClickListener {
                mFileSelectorRequest = REQUEST_CHOOSE_FILE
                mFileSelector = chooseFile(mFileSelectorRequest, this@FinChatFragment)
                second.dismiss()
            }
            first.tvCancel.setOnClickListener {
                second.dismiss()
            }
        }
    }
    
    override fun onSuccess(results: List<FileSelectResult>?) {
        if (results.isNullOrEmpty()) mActivity.infoToast("File tidak terpilih!")
        else {
            mSelectedPath.clear()
            mSelectedPath.addAll(results.map {
                it.filePath.toString()
            })
            when (mFileSelectorRequest) {
                REQUEST_CHOOSE_IMAGE -> prepareImage()
                REQUEST_CHOOSE_FILE -> sendFile()
            }
        }
    }

    override fun onError(e: Throwable?) {
        mActivity.infoToast(e?.message ?: "File tidak terpilih!")
    }
    
        @Suppress("DEPRECATION")
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        mFileSelector?.obtainResult(requestCode, resultCode, data)
        super.onActivityResult(requestCode, resultCode, data)
    }

}
```
![image](https://user-images.githubusercontent.com/48875981/108041822-a5c99900-7071-11eb-863b-e00600847c8b.png)
![image](https://user-images.githubusercontent.com/48875981/108041974-d1e51a00-7071-11eb-9024-5f9dc4573f08.png)

**Comment by javakam** [comment]
你可能使用了错误的`requestCode`数值 

You may have used the wrong `requestCode` value

😅
```
2021-02-16 17:55:33.261 12921-12921/? E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.ando.file.sample, PID: 12921
    java.lang.IllegalArgumentException: Can only use lower 16 bits for requestCode
        at androidx.fragment.app.FragmentActivity.checkForValidRequestCode(FragmentActivity.java:715)
        at androidx.fragment.app.FragmentActivity.startActivityFromFragment(FragmentActivity.java:794)
        ...
```

🍎  `androidx.fragment.app.FragmentActivity#checkForValidRequestCode`  source code : 

```
/**
 * Checks whether the given request code is a valid code by masking it with 0xffff0000. Throws
 * an {@link IllegalArgumentException} if the code is not valid.
 */
static void checkForValidRequestCode(int requestCode) {
    if ((requestCode & 0xffff0000) != 0) {
        throw new IllegalArgumentException("Can only use lower 16 bits for requestCode");
    }
}
```

**Comment by JosephSanjaya** [comment]
@javakam actually i already tempered with request code sir, for the first time i use 
```kotlin
companion object {
    private const val REQUEST_CHOOSE_IMAGE = 10
    private const val REQUEST_CHOOSE_FILE = 11
}
```
something like this, but the result is the same.

**Comment by JosephSanjaya** [comment]
![image](https://user-images.githubusercontent.com/48875981/108185166-995f4200-713e-11eb-8ba0-349a57c3170f.png)
![image](https://user-images.githubusercontent.com/48875981/108185236-abd97b80-713e-11eb-8b36-f5a78d3dc85d.png)

I think something is not right, i tried using the sample apps and it's also returning null on path and Unknown on FileType.
@javakam

**Comment by javakam** [comment]
🍎 这个`BUG`已经修复, 请升级到`v1.3.7` (This `BUG` has been fixed, please upgrade to `v1.3.7`)

①我已经成功`复现`了你描述的这个问题,  `Log`日志如下 : 

I have successfully `reproduced` the problem you described, and the `Log` log is as follows:

```
I/FileLogger: FileUri getPath -> Authority: com.android.providers.media.documents, Fragment: null, Port: -1, Query: null, Scheme: 
    content, Host: com.android.providers.media.documents, Segments: [document, document:153]
I/chatty: uid=10155(com.ando.file.sample) identical 4 lines
I/FileLogger: FileUri getPath -> Authority: com.android.providers.media.documents, Fragment: null, Port: -1, Query: null, Scheme: 
    content, Host: com.android.providers.media.documents, Segments: [document, document:153]
D/FileLogger: findFileType: UNKNOWN
I/FileLogger: filterUri: content://com.android.providers.media.documents/document/document%3A153 fileType=UNKNOWN 
   option=0 isOptionsNullOrEmpty=true
I/FileLogger: limitFileSize  : 18237 true
E/FileLogger: FileSelectCallBack onError File type mismatch
```

②推测你读取的文件是处在`公共目录`中 ( It is assumed that the file you read is in the `public directory`)

eg : `Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOCUMENTS)` :

![image](https://user-images.githubusercontent.com/15343661/108316453-e4895b80-71f7-11eb-88c3-cb6e70b487f1.png)

③`Android11` `公共目录`中的`非媒体文件`可以通过 [**SAF**](https://developer.android.com/guide/topics/providers/document-provider) 进行访问, 而通过传统的`ContentResolver`方式仍然无法获取`路径`. 不过`Uri`已经满足了绝大多数的需求, 因为你可以通过`Uri`进行操作,  如`上传文件`: 

`Non-media files` in `Android11` `Public Directory` can be accessed through [**SAF**](https://developer.android.com/guide/topics/providers/document-provider), however, the traditional `ContentResolver` method still cannot obtain the `path`. But `Uri` has met most of the needs, because you can use `Uri` to operate, Such as `upload files`:

```
val ins: InputStream? = context.contentResolver.openInputStream(uri)
```
or
```
val parcelFileDescriptor = context.contentResolver.openFileDescriptor(uri, "rw")
val fileDescriptor: FileDescriptor = parcelFileDescriptor!!.fileDescriptor
val fileInputStream :FileInputStream=FileInputStream(fileDescriptor)
```

目前, `路径`返回值是`null`, 当然也可以采用`复制`到其他位置后再读取新文件的路径, 可以参考:`ando.file.core.FileUri#getDownloadsPathQ`, 但是这种方式会占用用户的存储空间, 非常不推荐, 会在后续版本删除. 简而言之, `AndroidQ`及以后版本, 尽量使用`Uri`操作文件而不是`路径`

At present, `path` return value is `null`, of course, you can also use `copy` to another location and then read the path of the new file, you can refer to: `ando.file.core.FileUri#getDownloadsPathQ`, but this method will Occupies the user's storage space, very not recommended, will be deleted in subsequent versions. In short, `AndroidQ` and later versions, try to use `Uri` to manipulate files instead of `path`

④`Android` 目录结构 (`Android` directory structure)
```
Internal storage : /data/data/package/
External storage (private directory) : /storage/emulated/0/Android/data/xxx.xxx.xxx/
External storage (public directory) : /storage/emulated/0/
```

![image](https://user-images.githubusercontent.com/15343661/108338673-7e113700-7211-11eb-84b8-a634088e71c3.png)

`Google`建议在存储文件时, 最好放到`APP`的`私有目录`中 ( `Google` recommends storing files in the `private directory` of `APP` )

也许这个对你有所帮助 (Maybe this will help you) : 
English 👉 https://developer.android.com/about/versions/11/privacy/storage
中文版 👉 https://developer.android.com/about/versions/11/privacy/storage?hl=zh-cn

⑤Reference:
storage-samples 👉 https://github.com/android/storage-samples
Android 10 分区存储介绍及百度APP适配实践 👉 https://juejin.cn/post/6844904063432130568
共享存储空间-媒体内容 👉 https://developer.android.com/training/data-storage/shared/media
共享存储空间-文档和其他文件 👉 https://developer.android.com/training/data-storage/shared/documents-files

**Comment by JosephSanjaya** [comment]
@javakam Ok, sir understandable, thanks for your comprehensive explanation, it's working now.
